File: api/src/main/java/com/wbrawner/budgetserver/config/JdbcUserDetailsService.java
package com.wbrawner.budgetserver.config;

import com.wbrawner.budgetserver.user.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Component;

@Component
public class JdbcUserDetailsService implements UserDetailsService {

    private final UserRepository userRepository;

    @Autowired
    public JdbcUserDetailsService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        UserDetails userDetails;
        userDetails = userRepository.findByUsername(username).orElse(null);
        if (userDetails != null) {
            return userDetails;
        }
        userDetails = userRepository.findByEmail(username).orElse(null);
        if (userDetails != null) {
            return userDetails;
        }
        throw new UsernameNotFoundException("Unable to find user with username $username");
    }
}

File: api/src/main/kotlin/com/wbrawner/twigs/server/config/JdbcUserDetailsService.kt
package com.wbrawner.twigs.server.config

import com.wbrawner.budgetserver.user.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Component

@Component
open class JdbcUserDetailsService @Autowired constructor(private val userRepository: UserRepository) : UserDetailsService {
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        var userDetails: UserDetails?
        userDetails = userRepository.findByName(username).orElse(null)
        if (userDetails != null) {
            return userDetails
        }
        userDetails = userRepository.findByEmail(username).orElse(null)
        if (userDetails != null) {
            return userDetails
        }
        throw UsernameNotFoundException("Unable to find user with username \$username")
    }
}